polynote-kernel/src/main/scala/polynote/kernel/util/TaskManager.scala

```diff
@@ -1,14 +1,160 @@
 package polynote.kernel.util
 
-import cats.effect.IO
-import cats.effect.concurrent.Deferred
+import java.util.concurrent.{ConcurrentHashMap, ConcurrentLinkedDeque, LinkedBlockingQueue}
 
-// TODO: trying to centralize how tasks are tracked.
-class TaskManager {
+import scala.collection.JavaConverters._
+import cats.syntax.all._
+import cats.instances.list._
+import cats.effect.concurrent.{Deferred, Ref, Semaphore}
+import cats.effect.{ContextShift, Fiber, IO}
+import fs2.Stream
+import fs2.concurrent.Queue
+import polynote.kernel.{KernelStatusUpdate, TaskInfo, TaskStatus, UpdatedTasks}
 
-  def add(id: String, cancel: () => IO[Unit]): Unit = {
+/**
+  * Manages running tasks. Announces task status to `statusUpdates`. Tasks can be run concurrently or queued to run
+  * one at a time.
+  */
+class TaskManager(
+  statusUpdates: Publish[IO, KernelStatusUpdate],
+  semaphore: Semaphore[IO])(implicit
+  contextShift: ContextShift[IO]
+) {
+  import TaskManager.{Queued, Running, QueuedStream}
 
+  private val running = new ConcurrentLinkedDeque[TaskManager.Running[_]]()
+  private val queuedTasks = new LinkedBlockingQueue[TaskManager.Queued[_]]()
+
+  private def publish(taskInfo: TaskInfo): IO[Unit] = statusUpdates.publish1(UpdatedTasks(taskInfo :: Nil))
+
+  private def run[A](taskInfo: TaskInfo, task: IO[A]): IO[Running[A]] = {
+    for {
+      _       <- publish(taskInfo)
+      fiber   <- task.start
+      runTask  = Running(taskInfo, fiber)
+      _       <- IO(running.add(runTask))
+    } yield runTask
+  }
+
+  private def runNext(): IO[Unit] = (IO(queuedTasks.take()) <* semaphore.acquire).flatMap {
+    queued =>
+      System.err.println(s"Running ${queued.taskInfo.label}")
+      run(queued.taskInfo.running, queued.io).flatMap {
+        runTask =>
+          def finalizer: IO[Unit] =
+            IO(running.remove(runTask)) *>
+              publish(queued.taskInfo.completed) *>
+                (runNext().start >>= runQueueFiber.set) *>
+                  semaphore.release
+
+          runTask.fiber.join.uncancelable.flatMap {
+            result =>
+              queued.onComplete(result, finalizer)
+          }.attempt.flatMap {
+            result =>
+              queued.deferred.complete(result) *> {
+                result match {
+                  case Left(err) => finalizer
+                  case Right(_)  => IO.unit
+                }
+              }
+          }
```

Comment:
This API seems a little odd... something about passing `result` and `finalizer` to `onComplete` and then passing `result` again to `complete` and then at the end of the day returning either an empty Unit or `finalizer` ... 

I think I understand why it's like this and I'm not sure I can think of a better way to do it, but it does seem a little strange...?

Reply:
Agreed, I'm not super happy with it. The reason is to unify the existential type to complete the deferred. The deferred song and dance is because the task queue is going to handle actually evaluating the `IO`, so we can't just return that `IO`. Will try to rework `TaskManager` some more in a future PR.